tests/frontend/test_program.py

```diff
@@ -650,6 +650,37 @@ class DummyCircuit(Compiler):
                 compiler=DummyCircuit(),
             )
 
+    def test_missing_layout(self):
+        """Test that an error is raised if the device spec is missing a layout when compiled"""
+        device_dict = {
+            "target": None,
+            "layout": None,
+            "modes": 2,
+            "compiler": ["DummyCompiler"],
+            "gate_parameters": {
```

Comment:
Maybe there should now also be a test where the `device_dict` does not contain any `gate_parameters`? I'm somewhat curious if it makes sense, even for a simulator, to have a layout but no gate parameters...

Reply:
I can add this. Currently, the layout isn't used at all if the spec doesn't contain any `gate_parameters` (both if it's an empty dict or `None`), unless using a `TDMProgram`, in which case an error would be raised.